### ISSUE TITLE
Make OpenSSL use pointers from input buffer for holding CMS structure instead of allocating memory

### ISSUE THREAD

**Issue by mura1i** [openssl/openssl#24832]
### Discussed in https://github.com/openssl/openssl/discussions/24822

<div type='discussions-op-text'>

<sup>Originally posted by **mura1i** July  9, 2024</sup>
An embedded device with limited memory is already using OpenSSL for communication.

I am implementing a CMS signature verification wrapper library using OpenSSL that should work for the device.

We already have the below code which is working.

```c
CMS_ContentInfo *cms = d2i_CMS_ContentInfo(nullptr, in, len); // allocates memory 1

if (CMS_verify(cms, nullptr, store, nullptr, nullptr, 0))
{
    ASN1_OCTET_STRING **data = CMS_get0_content(cms);
    len = (*data)->length);
    out = OPENSSL_zalloc(len); // allocates memory 2
    memcpy(out, (*data)->data, len);
}

CMS_ContentInfo_free(cms); // frees memory 1
```

But it allocates redundant memory 2 times.

Any way to make OpenSSL use the input buffer itself for holding the data?</div>

Found out that this is not currently feasible from the discussion. Raising this as a feature request hoping someone would find a way in future 🤞

**Comment by t8m** [comment]
This is unlikely to get priority to be fixed by the OpenSSL team itself. Marking as community issue with help wanted.

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 2 months otherwise.

**Comment by DDvO** [comment]
@mura1i looks like you do not really need to copy `(*data)->data` over to a new section of memory.
Why not simply retain the `CMS_ContentInfo` structure (i.e., not yet use `CMS_ContentInfo_free(cms)`)
as long as you need access to `(*data)->data`, and use that existing pointer instead?

**Comment by mura1i** [comment]
@DDvO that is true, but if this feature is implemented, both allocations won't be needed.

**Comment by DDvO** [comment]
Good that you clarified this - I found your application example confusing/misleading.

So you are after an alternative to `d2i_CMS_ContentInfo()` that does des DER decoding
but in the signedData case allocates fresh memory only for the CMS structures themselves while leaving the included pure payload data where it is (as part of the `in` buffer), and instead just point to it.

I fear that that will be hard to do because the `d2i_*` functions use generic code for decoding any DER-encoded structures, 
which so far freshly allocates memory for all included `OCTET STRING` data.
Using just an aliased pointer to the respective section of the input buffer would also cause a non-standard ownership relation: the input buffer must only be deallocated along with the resulting `cms` pointer.

**Comment by t8m** [comment]
This issue was closed because it was marked with the help wanted label and there was no interest in implementing this issue or any other activity leading to implementing or re-prioritizing it for more than 2 months.
